File: src/transport/transport.ts
/**
 * A2UI WebSocket Transport Layer
 * Framework-agnostic agent communication
 */

import type {
  A2UIMessage,
  CreateSurfaceMessage,
  UpdateComponentsMessage,
  UpdateDataModelMessage,
  DeleteSurfaceMessage,
  UserActionMessage,
  ErrorMessage,
  SearchVideosMessage,
  SearchResultsMessage,
} from '../types/index.js'

/**
 * Transport connection status
 */
export type TransportStatus = 'disconnected' | 'connecting' | 'connected' | 'error'

/**
 * Transport options
 */
export interface TransportOptions {
  /** Auto-reconnect on connection loss */
  autoReconnect?: boolean
  /** Reconnect delay in ms */
  reconnectDelay?: number
  /** Max reconnect attempts (0 = infinite) */
  maxReconnectAttempts?: number
  /** Ping interval in ms (0 = disabled) */
  pingInterval?: number
  /** Pong timeout in ms */
  pongTimeout?: number
}

/**
 * Event handler type
 */
export type EventHandler<T = unknown> = (data: T) => void

/**
 * A2UI Transport class
 * Manages WebSocket connection and message routing
 */
export class A2UITransport {
  private ws: WebSocket | null = null
  private _status: TransportStatus = 'disconnected'
  private readonly url: string
  private readonly options: Required<TransportOptions>
  private reconnectAttempts = 0
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null
  private pingTimer: ReturnType<typeof setInterval> | null = null
  private pongTimer: ReturnType<typeof setTimeout> | null = null

  // Event handlers
  private readonly handlers = new Map<string, Set<EventHandler>>()

  constructor(url: string, options: TransportOptions = {}) {
    this.url = url
    this.options = {
      autoReconnect: options.autoReconnect ?? true,
      reconnectDelay: options.reconnectDelay ?? 3000,
      maxReconnectAttempts: options.maxReconnectAttempts ?? 5,
      pingInterval: options.pingInterval ?? 30000,
      pongTimeout: options.pongTimeout ?? 5000,
    }
  }

  /**
   * Connect to the agent
   */
  async connect(): Promise<void> {
    if (this._status === 'connected' || this._status === 'connecting') {
      return
    }

    this._status = 'connecting'
    this.emit('statusChange', this._status)

    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.url)

        this.ws.onopen = () => {
          this._status = 'connected'
          this.reconnectAttempts = 0
          this.emit('statusChange', this._status)
          this.emit('connect', undefined)
          this.startPing()
          resolve()
        }

        this.ws.onmessage = (event) => {
          this.handleMessage(event.data as string)
        }

        this.ws.onerror = () => {
          this._status = 'error'
          this.emit('statusChange', this._status)
          this.emit('error', new Error('WebSocket error'))
          reject(new Error('WebSocket connection failed'))
        }

        this.ws.onclose = () => {
          this.handleClose()
        }
      } catch (error) {
        this._status = 'error'
        this.emit('statusChange', this._status)
        reject(error)
      }
    })
  }

  /**
   * Disconnect from the agent
   */
  disconnect(): void {
    this.stopPing()
    this.stopReconnect()

    if (this.ws) {
      this.ws.close()
      this.ws = null
    }

    this._status = 'disconnected'
    this.emit('statusChange', this._status)
    this.emit('disconnect', undefined)
  }

  /**
   * Send a message to the agent
   */
  send(message: A2UIMessage): void {
    if (!this.isConnected) {
      throw new Error('Cannot send message: not connected')
    }

    const json = JSON.stringify(message)
    this.ws!.send(json)
  }

  /**
   * Register an event handler
   */
  on(event: 'statusChange', handler: EventHandler<TransportStatus>): void
  on(event: 'connect', handler: EventHandler<undefined>): void
  on(event: 'disconnect', handler: EventHandler<undefined>): void
  on(event: 'reconnecting', handler: EventHandler<number>): void
  on(event: 'error', handler: EventHandler<Error>): void
  on(event: 'message', handler: EventHandler<A2UIMessage>): void
  on(event: 'createSurface', handler: EventHandler<CreateSurfaceMessage>): void
  on(event: 'updateComponents', handler: EventHandler<UpdateComponentsMessage>): void
  on(event: 'updateDataModel', handler: EventHandler<UpdateDataModelMessage>): void
  on(event: 'deleteSurface', handler: EventHandler<DeleteSurfaceMessage>): void
  on(event: 'userAction', handler: EventHandler<UserActionMessage>): void
  on(event: 'searchVideos', handler: EventHandler<SearchVideosMessage>): void
  on(event: 'searchResults', handler: EventHandler<SearchResultsMessage>): void
  on<T = unknown>(event: string, handler: EventHandler<T>): void
  on<T = unknown>(event: string, handler: EventHandler<T>): void {
    if (!this.handlers.has(event)) {
      this.handlers.set(event, new Set())
    }
    this.handlers.get(event)!.add(handler as EventHandler)
  }

  /**
   * Unregister an event handler
   */
  off(event: 'statusChange', handler: EventHandler<TransportStatus>): void
  off(event: 'connect', handler: EventHandler<undefined>): void
  off(event: 'disconnect', handler: EventHandler<undefined>): void
  off(event: 'reconnecting', handler: EventHandler<number>): void
  off(event: 'error', handler: EventHandler<Error>): void
  off(event: 'message', handler: EventHandler<A2UIMessage>): void
  off(event: 'createSurface', handler: EventHandler<CreateSurfaceMessage>): void
  off(event: 'updateComponents', handler: EventHandler<UpdateComponentsMessage>): void
  off(event: 'updateDataModel', handler: EventHandler<UpdateDataModelMessage>): void
  off(event: 'deleteSurface', handler: EventHandler<DeleteSurfaceMessage>): void
  off(event: 'userAction', handler: EventHandler<UserActionMessage>): void
  off(event: 'searchVideos', handler: EventHandler<SearchVideosMessage>): void
  off(event: 'searchResults', handler: EventHandler<SearchResultsMessage>): void
  off<T = unknown>(event: string, handler: EventHandler<T>): void
  off<T = unknown>(event: string, handler: EventHandler<T>): void {
    const handlers = this.handlers.get(event)
    if (handlers) {
      handlers.delete(handler as EventHandler)
    }
  }

  /**
   * Get current status
   */
  get status(): TransportStatus {
    return this._status
  }

  /**
   * Check if connected
   */
  get isConnected(): boolean {
    return this._status === 'connected' && this.ws !== null && this.ws.readyState === WebSocket.OPEN
  }

  /**
   * Handle incoming message
   */
  private handleMessage(data: string): void {
    try {
      const message = JSON.parse(data) as A2UIMessage

      // Handle pong response
      if (message.type === 'pong') {
        this.handlePong()
        return
      }

      // Emit message to specific type handlers
      this.emit(message.type, message)

      // Emit to generic message handler
      this.emit('message', message)
    } catch (error) {
      this.emit('error', new Error(`Failed to parse message: ${String(error)}`))
    }
  }

  /**
   * Handle connection close
   */
  private handleClose(): void {
    this.stopPing()

    const wasConnected = this._status === 'connected'
    this._status = 'disconnected'
    this.emit('statusChange', this._status)

    if (wasConnected) {
      this.emit('disconnect', undefined)

      // Auto-reconnect if enabled
      if (this.options.autoReconnect) {
        this.scheduleReconnect()
      }
    }
  }

  /**
   * Schedule reconnection attempt
   */
  private scheduleReconnect(): void {
    if (this.reconnectAttempts >= this.options.maxReconnectAttempts && this.options.maxReconnectAttempts > 0) {
      this.emit('error', new Error('Max reconnect attempts reached'))
      return
    }

    this.reconnectAttempts++
    this.emit('reconnecting', this.reconnectAttempts)

    this.reconnectTimer = setTimeout(() => {
      this.connect().catch((error) => {
        this.emit('error', error as Error)
      })
    }, this.options.reconnectDelay)
  }

  /**
   * Stop reconnection timer
   */
  private stopReconnect(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer)
      this.reconnectTimer = null
    }
    this.reconnectAttempts = 0
  }

  /**
   * Start ping/pong keep-alive
   */
  private startPing(): void {
    if (this.options.pingInterval === 0) {
      return
    }

    this.pingTimer = setInterval(() => {
      if (this.isConnected) {
        this.send({ type: 'ping' })
        this.schedulePongTimeout()
      }
    }, this.options.pingInterval)
  }

  /**
   * Stop ping/pong
   */
  private stopPing(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer)
      this.pingTimer = null
    }
    if (this.pongTimer) {
      clearTimeout(this.pongTimer)
      this.pongTimer = null
    }
  }

  /**
   * Schedule pong timeout
   */
  private schedulePongTimeout(): void {
    if (this.pongTimer) {
      clearTimeout(this.pongTimer)
    }

    this.pongTimer = setTimeout(() => {
      this.emit('error', new Error('Pong timeout: connection may be dead'))
      this.ws?.close()
    }, this.options.pongTimeout)
  }

  /**
   * Handle pong response
   */
  private handlePong(): void {
    if (this.pongTimer) {
      clearTimeout(this.pongTimer)
      this.pongTimer = null
    }
  }

  /**
   * Emit event to handlers
   */
  private emit<T = unknown>(event: string, data: T): void {
    const handlers = this.handlers.get(event)
    if (handlers) {
      handlers.forEach((handler) => handler(data))
    }
  }
}

File: sdks/go/a2ui/transport.go
package a2ui

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// TransportStatus represents the connection status
type TransportStatus string

const (
	// StatusDisconnected means not connected
	StatusDisconnected TransportStatus = "disconnected"
	// StatusConnecting means connection in progress
	StatusConnecting TransportStatus = "connecting"
	// StatusConnected means connected and ready
	StatusConnected TransportStatus = "connected"
	// StatusError means error state
	StatusError TransportStatus = "error"
)

// TransportOptions configures the transport behavior
type TransportOptions struct {
	// AutoReconnect enables automatic reconnection on connection loss
	AutoReconnect bool
	// ReconnectDelay is the delay between reconnection attempts (default: 3s)
	ReconnectDelay time.Duration
	// MaxReconnectAttempts limits reconnection attempts (0 = infinite)
	MaxReconnectAttempts int
	// PingInterval is the interval between ping messages (0 = disabled)
	PingInterval time.Duration
	// PongTimeout is the timeout for pong responses (default: 5s)
	PongTimeout time.Duration
	// Logger for transport events (optional)
	Logger *log.Logger
}

// DefaultTransportOptions returns default transport options
func DefaultTransportOptions() *TransportOptions {
	return &TransportOptions{
		AutoReconnect:        true,
		ReconnectDelay:       3 * time.Second,
		MaxReconnectAttempts: 5,
		PingInterval:         30 * time.Second,
		PongTimeout:          5 * time.Second,
	}
}

// MessageHandler is a function that handles A2UI messages
type MessageHandler func(message json.RawMessage) error

// EventHandler is a function that handles transport events
type EventHandler func(data interface{})

// Transport handles WebSocket communication with A2UI agents
type Transport struct {
	url     string
	options *TransportOptions

	// Connection state
	conn              *websocket.Conn
	status            TransportStatus
	reconnectAttempts int
	mu                sync.RWMutex
	ctx               context.Context
	cancel            context.CancelFunc

	// Message handlers (by message type)
	handlers   map[string][]MessageHandler
	handlersMu sync.RWMutex

	// Event handlers
	eventHandlers   map[string][]EventHandler
	eventHandlersMu sync.RWMutex

	// Timers
	pingTimer *time.Timer
	pongTimer *time.Timer
}

// NewTransport creates a new A2UI transport
func NewTransport(url string, options *TransportOptions) *Transport {
	if options == nil {
		options = DefaultTransportOptions()
	}

	ctx, cancel := context.WithCancel(context.Background())

	return &Transport{
		url:           url,
		options:       options,
		status:        StatusDisconnected,
		handlers:      make(map[string][]MessageHandler),
		eventHandlers: make(map[string][]EventHandler),
		ctx:           ctx,
		cancel:        cancel,
	}
}

// Connect establishes the WebSocket connection
func (t *Transport) Connect() error {
	t.mu.Lock()
	if t.status == StatusConnected || t.status == StatusConnecting {
		t.mu.Unlock()
		return nil
	}
	t.status = StatusConnecting
	t.mu.Unlock()

	t.emit("statusChange", t.status)

	dialer := websocket.DefaultDialer
	conn, _, err := dialer.Dial(t.url, nil)
	if err != nil {
		t.mu.Lock()
		t.status = StatusError
		t.mu.Unlock()
		t.emit("statusChange", t.status)
		t.emit("error", err)
		return fmt.Errorf("failed to connect: %w", err)
	}

	t.mu.Lock()
	t.conn = conn
	t.status = StatusConnected
	t.reconnectAttempts = 0
	t.mu.Unlock()

	t.emit("statusChange", t.status)
	t.emit("connect", nil)

	// Start message receiver
	go t.receiveLoop()

	// Start ping/pong if enabled
	if t.options.PingInterval > 0 {
		t.startPing()
	}

	t.logf("Connected to %s", t.url)
	return nil
}

// Disconnect closes the WebSocket connection
func (t *Transport) Disconnect() error {
	t.mu.Lock()
	defer t.mu.Unlock()

	if t.conn != nil {
		err := t.conn.Close()
		t.conn = nil
		t.status = StatusDisconnected
		t.emit("statusChange", t.status)
		t.emit("disconnect", nil)
		t.logf("Disconnected from %s", t.url)
		return err
	}

	return nil
}

// Close permanently closes the transport and cancels all operations
func (t *Transport) Close() error {
	t.cancel()
	return t.Disconnect()
}

// Send sends a message to the agent
func (t *Transport) Send(message Message) error {
	t.mu.RLock()
	conn := t.conn
	status := t.status
	t.mu.RUnlock()

	if status != StatusConnected || conn == nil {
		return errors.New("not connected")
	}

	data, err := json.Marshal(message)
	if err != nil {
		return fmt.Errorf("failed to marshal message: %w", err)
	}

	err = conn.WriteMessage(websocket.TextMessage, data)
	if err != nil {
		return fmt.Errorf("failed to send message: %w", err)
	}

	return nil
}

// On registers a message handler for a specific message type
func (t *Transport) On(messageType string, handler MessageHandler) {
	t.handlersMu.Lock()
	defer t.handlersMu.Unlock()

	t.handlers[messageType] = append(t.handlers[messageType], handler)
}

// OnEvent registers an event handler
// Events: "connect", "disconnect", "error", "statusChange", "reconnecting"
func (t *Transport) OnEvent(event string, handler EventHandler) {
	t.eventHandlersMu.Lock()
	defer t.eventHandlersMu.Unlock()

	t.eventHandlers[event] = append(t.eventHandlers[event], handler)
}

// Status returns the current connection status
func (t *Transport) Status() TransportStatus {
	t.mu.RLock()
	defer t.mu.RUnlock()
	return t.status
}

// IsConnected returns true if connected
func (t *Transport) IsConnected() bool {
	return t.Status() == StatusConnected
}

// receiveLoop handles incoming messages
func (t *Transport) receiveLoop() {
	for {
		select {
		case <-t.ctx.Done():
			return
		default:
			t.mu.RLock()
			conn := t.conn
			t.mu.RUnlock()

			if conn == nil {
				return
			}

			_, message, err := conn.ReadMessage()
			if err != nil {
				t.logf("Read error: %v", err)
				t.handleDisconnect()
				return
			}

			// Parse message type
			var baseMsg struct {
				Type string `json:"type"`
			}
			if err := json.Unmarshal(message, &baseMsg); err != nil {
				t.emit("error", fmt.Errorf("failed to parse message: %w", err))
				continue
			}

			// Handle pong specially
			if baseMsg.Type == "pong" {
				t.handlePong()
				continue
			}

			// Call registered handlers
			t.handlersMu.RLock()
			handlers := t.handlers[baseMsg.Type]
			t.handlersMu.RUnlock()

			for _, handler := range handlers {
				if err := handler(json.RawMessage(message)); err != nil {
					t.emit("error", fmt.Errorf("handler error for %s: %w", baseMsg.Type, err))
				}
			}

			// Also emit to generic "message" handlers
			t.handlersMu.RLock()
			messageHandlers := t.handlers["message"]
			t.handlersMu.RUnlock()

			for _, handler := range messageHandlers {
				if err := handler(json.RawMessage(message)); err != nil {
					t.emit("error", fmt.Errorf("message handler error: %w", err))
				}
			}
		}
	}
}

// handleDisconnect handles connection loss
func (t *Transport) handleDisconnect() {
	t.mu.Lock()
	wasConnected := t.status == StatusConnected
	t.status = StatusDisconnected
	t.mu.Unlock()

	t.emit("statusChange", t.status)

	if wasConnected {
		t.emit("disconnect", nil)
		t.logf("Connection lost")

		// Auto-reconnect if enabled
		if t.options.AutoReconnect {
			t.scheduleReconnect()
		}
	}
}

// scheduleReconnect schedules a reconnection attempt
func (t *Transport) scheduleReconnect() {
	t.mu.Lock()
	if t.options.MaxReconnectAttempts > 0 &&
		t.reconnectAttempts >= t.options.MaxReconnectAttempts {
		t.mu.Unlock()
		t.emit("error", errors.New("max reconnect attempts reached"))
		return
	}

	t.reconnectAttempts++
	attempt := t.reconnectAttempts
	t.mu.Unlock()

	t.emit("reconnecting", attempt)
	t.logf("Reconnecting (attempt %d)...", attempt)

	time.AfterFunc(t.options.ReconnectDelay, func() {
		if err := t.Connect(); err != nil {
			t.emit("error", fmt.Errorf("reconnect failed: %w", err))
		}
	})
}

// startPing starts the ping/pong keep-alive mechanism
func (t *Transport) startPing() {
	if t.options.PingInterval <= 0 {
		return
	}

	t.pingTimer = time.AfterFunc(t.options.PingInterval, func() {
		if t.IsConnected() {
			// Send ping
			ping := &PingMessage{Type: MessageTypePing}
			if err := t.Send(ping); err != nil {
				t.emit("error", fmt.Errorf("failed to send ping: %w", err))
			}

			// Schedule pong timeout
			t.pongTimer = time.AfterFunc(t.options.PongTimeout, func() {
				t.emit("error", errors.New("pong timeout: connection may be dead"))
				t.handleDisconnect()
			})

			// Schedule next ping
			t.startPing()
		}
	})
}

// handlePong handles pong response
func (t *Transport) handlePong() {
	if t.pongTimer != nil {
		t.pongTimer.Stop()
		t.pongTimer = nil
	}
}

// emit emits an event to registered handlers
func (t *Transport) emit(event string, data interface{}) {
	t.eventHandlersMu.RLock()
	handlers := t.eventHandlers[event]
	t.eventHandlersMu.RUnlock()

	for _, handler := range handlers {
		handler(data)
	}
}

// logf logs a message if logger is configured
func (t *Transport) logf(format string, args ...interface{}) {
	if t.options.Logger != nil {
		t.options.Logger.Printf(format, args...)
	}
}
